pkg/controller/sdk/provider.go

```diff
@@ -29,8 +28,6 @@ import (
 
 type KnativeReconciler interface {
 	Reconcile(ctx context.Context, object runtime.Object) (runtime.Object, error)
-	InjectClient(c client.Client) error
```

Comment:
I'm surprised this line can be removed. Are any of these reconcilers viable without a `client`?

Reply:
The `inject.ClientInto` method will detect if the reconciler implements `InjectClient` and call it if so.

There's no need to require this method in the interface because there are other ways of injecting the client. For example, in `pkg/controller/containersource/provider.go`:

```diff
   p := &sdk.Provider{
     AgentName: controllerAgentName,
     Parent:    &v1alpha1.ContainerSource{},
     Owns:      []runtime.Object{&appsv1.Deployment{}},
     Reconciler: &reconciler{
       recorder: mgr.GetRecorder(controllerAgentName),
       scheme:   mgr.GetScheme(),
+      client:  mgr.GetClient(),
     },
   }
```

Reply:
So my preference is to continue to require it in the interface. You are right that there are other ways to do it, but do they provide any advantages?

If an author forgets to somehow inject the client (via whichever method they prefer), then this controller will fail at runtime. And even then, only after the controller is trying to do something. I'm a fan of keeping this a compile-time error 😄